packages/@aws-cdk/cdk/lib/core/tag-manager.ts

```diff
@@ -0,0 +1,149 @@
+ */
+export interface ITaggable {
+    tags: TagManager,
+}
+
+/**
+ * A single tag property values
+ */
+interface ITagProps {
+    value: string | Token;
+    propagate?: boolean;
+}
+
+/**
+ * A an object of tags
+ */
+export interface ITags {
+    [key: string]: ITagProps;
+}
+
+/**
+ * A single tag
+ */
+export interface ITag {
+    key: string;
+    value: string | Token;
+}
+
+/**
+ * TagManager facilitates a common implementation of tagging for Constructs.
+ *
+ * Each construct that wants to support tags should implement the `ITaggable`
+ * interface and properly pass tags to the `Resources` (Cloudformation) elements
+ * the `Construct` creates using `toToken()` for lazy evaluations
+ */
+export class TagManager {
+
+    /**
+     * Checks if the object implements the `ITaggable` interface
+     */
+    public static isTaggable(taggable: ITaggable | any): taggable is ITaggable {
+        return ((taggable as ITaggable).tags !== undefined);
+    }
+
+    private readonly _tags: ITags = {};
+
+    private readonly parentTagManager: TagManager | undefined;
+
+    constructor(partner: Construct, initialTags: ITags = {}) {
+        if (TagManager.isTaggable(partner.parent)) {
+            this.parentTagManager = partner.parent.tags;
+        } else {
+            this.parentTagManager = undefined;
+        }
+        this._tags = this.defaultPropagate(initialTags);
+    }
+
+    /**
+     * Returns a deep copy of the tags object, except for nested Tokens
+     *
+     * Deep copies of all primiatve `ITagProps` are returned. If a `value` is a
+     * `Token` then that reference is kept and not deep copied. The assumption is
+     * this is for lazy evaluation. Modifying a nested `Token` is not advised
+     * the consumer should use `addTags` with the corresponding key to change
+     * the value or override a propagated parent tag.
+     */
+    public get tags(): ITags {
+        const tags: ITags = {};
+        Object.keys(this._tags).map( key => (tags[key] = {...this._tags[key]}));
+        return {...tags, ...this.propagatedTags()};
+    }
+
+    /**
+     * Converts the `tags` to a Token for use in lazy evaluation
+     */
+    public toToken(propagateOnly = false): Token {
+        return new Token(() => this.toArray(propagateOnly));
+    }
+
+    /**
+     * Returns a deep copy of the tags as `Array<{key: string, value: any}>`
+     */
+    public toArray(propagateOnly = false): ITag[] {
+        if (propagateOnly) {
+            const propTags = this.propagatedTags();
+            return Object.keys(propTags).map( key => ({key, value: propTags[key].value}));
+        }
+        const tags = this.tags;
+        return Object.keys(tags).map( key => ({key, value: tags[key].value}));
+    }
+
+    /**
+     * Adds the specified tag to the array of tags
+     *
+     * @param tag The tag to add
+     * @param propagate  Whether to add this tag to all children
+     */
+    public addTags(tags: ITags): void {
```

Comment:
Ergonomics: the common use case would be to just set a propagating tag. And `key, value` would be the intuition of most users when they set a tag:

```ts
foo.setTag(key, value);
```

Then, we can include `options: TagOptions = {}`:

```ts
foo.setTag('key', 'value', { propagate: false, overwrite: true })
```

Reply:
Also transitioned to `setTag` because `addTag` implies an always adding and this will overwrite. I'm a little nervous because JSII seems to prevent `getXXX` for Java name collision protection, but not `setXXX`. If I need to change this let me know.